### PR TITLE
Feature/serialtube

### PR DESCRIPTION
**Pull request by JonathanBeverley**
Add a class to handle serial connections.
There is a significant bugfix in `lib/pwnlib/tubes/tube.rb`, which is necessary for it to work properly.
I've bumped version to 1.0.2 (mostly so I could tell I was using it in testing.)

---
This change is [Reviewable](https://reviewable.io/reviews/peter50216/pwntools-ruby/90)

### CODE REVIEW

**Comment by david942j** [comment]
Thanks for distributing!

But please fixed these issues before we start to review:
* Don't bump the version, you can do it locally but don't commit it
* Resolve conflicts
* Add tests and ensure 100% coverage
    - Beware of our tests run under three operating systems
* Add documents
* Follow STYLE.md, some issues I found in a quick review:
    - Use first instead of [0]
    - Sort the requiring
    - Sort the dependency in gemspec

Thanks again!

**Comment by david942j** [comment]
And it would be great if you file an issue to describe how the bug effects in tube.rb

File: lib/pwnlib/timer.rb [hunk 1]
```diff
@@ -54,8 +54,9 @@ def countdown(timeout = nil)
       begin
         yield
       ensure
-        raise ::Pwnlib::Errors::TimeoutError unless active?
+        was_active = active?
```
Comment (david942j) [inline lib/pwnlib/timer.rb]: :+1:

File: lib/pwnlib/tubes/tube.rb [hunk 2]
```diff
@@ -159,7 +166,7 @@ def recvuntil(delims, drop: false, timeout: nil)
             while @timer.active?
               s = recv(1)
 
-              return '' if s.empty?
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: Is this change needed?

File: lib/pwnlib/tubes/tube.rb [hunk 3]
```diff
@@ -101,13 +104,17 @@ def recvpred(timeout: nil)
 
               c = recv(1)
 
-              return '' if c.empty?
+              next if c.empty?
               data << c
             end
             data.slice!(0..-1)
-          ensure
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: Why this?
Original code seems correct

File: test/tubes/serialtube_test.rb [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+require 'open3'
+
+require 'test_helper'
+
+require 'pwnlib/tubes/serialtube'
+
+class SerialTest < MiniTest::Test
+  include ::Pwnlib::Tubes
+
```
Comment (david942j) [inline test/tubes/serialtube_test.rb]: Is the serial tube intend to work on macOS?
If yes, let the tests pass on macOS (install socat)
If no, skip the test unless the environment is on Linux

Comment (david942j) [reply test/tubes/serialtube_test.rb]: And skip on Windows as well

File: lib/pwnlib/pwn.rb [hunk 5]
```diff
@@ -13,6 +13,7 @@
 require 'pwnlib/reg_sort'
 require 'pwnlib/shellcraft/shellcraft'
 require 'pwnlib/tubes/sock'
+require 'pwnlib/tubes/serialtube'
```
Comment (david942j) [inline lib/pwnlib/pwn.rb]: Put this line before 'pwnlib/tubes/sock'

File: lib/pwnlib/tubes/tube.rb (hunk 3 above)
Comment (JonathanBeverley) [reply lib/pwnlib/tubes/tube.rb]: This might be a disagreement on what the function is supposed to do. In my opinion, it should not return until the timeout expires unless it has a match with the predicate. Just because a particular recv(1) didn't return anything doesn't mean that one .5s later won't.

Swap from rescue->raise is different. That's because we must not unrecv(data) if we match the predicate. With ensure, we always unrecv.

File: lib/pwnlib/tubes/tube.rb (hunk 2 above)
Comment (JonathanBeverley) [reply lib/pwnlib/tubes/tube.rb]: No. I raised it in issue #91, and it was classified as "notbug".

File: lib/pwnlib/timer.rb (hunk 1 above)
Comment (david942j) [reply lib/pwnlib/timer.rb]: This bug is fixed by #116

File: test/tubes/serialtube_test.rb (hunk 4 above)
Comment (david942j) [reply test/tubes/serialtube_test.rb]: To install socat, simply add `brew install socat` in travis/install.sh#setup_osx

Is that possible to test serialtube on Windows? I said skip on Windows because the tests you added (`socat ...`) can't be supported on Windows.

If serialtube is supposed to support on Windows, maybe you should try another testing method to make it work on Windows as well.
Otherwise, just forget Windows supports and skip the tests.

File: test/tubes/serialtube_test.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  def skip_windows
+    skip 'Not test tube/serialtube on Windows' if TTY::Platform.new.windows?
+  end
+
+  def open_pair
+    Open3.popen3('socat -d -d pty,raw,echo=0 pty,raw,echo=0') do |_i, _o, stderr, thread|
+      devs = []
+      2.times do
+        devs << stderr.readline.chomp.split.last
+        raise IOError, 'Could not create serial crosslink' if devs.last !~ %r{^/dev/pts/[0-9]+$}
```
Comment (david942j) [inline test/tubes/serialtube_test.rb]: I tested on macOS, the output of socat is
```
➔ socat -d -d pty,raw,echo=0 pty,raw,echo=0
2018/06/10 13:44:41 socat[24931] N PTY is /dev/ttys005
2018/06/10 13:44:41 socat[24931] N PTY is /dev/ttys006
2018/06/10 13:44:41 socat[24931] N starting data transfer loop with FDs [5,5] and [8,8]
```
maybe that's the reason tests fails on travis-CI

Comment (JonathanBeverley) [reply test/tubes/serialtube_test.rb]: You nailed it. I had a strict regex filter that matched Linux style PTYs.
Thanks.

File: lib/pwnlib/tubes/serialtube.rb [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      #
+      # @param [Integer] numbytes
+      #   An upper limit on the number of bytes to get.
+      #
+      # @return [String]
+      #   A string containing read bytes.
+      #
+      # @!macro raise_eof
+      def recv_raw(numbytes)
+        raise EOFError if @conn.nil?
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: `raise ::Pwnlib::Errors::EndOfTubeError`

File: lib/pwnlib/tubes/serialtube.rb [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            while @serial_timer.active?
+              data += @conn.read(numbytes - data.length)
+              break unless data.empty?
+              sleep 0.1
+            end
+            # TODO: should we reverse @convert_newlines here?
+            return data
+          rescue RubySerial::Error
+            close
+            raise EOFError
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: Same here

File: lib/pwnlib/tubes/serialtube.rb [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+      #
+      # @param [String] data
+      #   A string of the bytes to send.
+      #
+      # @return [Integer]
+      #   The number of bytes successfully written.
+      #
+      # @!macro raise_eof
+      def send_raw(data)
+        raise EOFError if @conn.nil?
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: Same

File: lib/pwnlib/tubes/serialtube.rb [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+        data.gsub!(context.newline, "\r\n") if @convert_newlines
+        begin
+          return @conn.write(data)
+        rescue RubySerial::Error
+          close
+          raise EOFError
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: Here

File: lib/pwnlib/tubes/serialtube.rb [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        end
+      end
+
+      # Sets the +timeout+ to use for subsequent +recv_raw+ calls.
+      #
+      # @param [Integer] timeout
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: The type is `Float?`
And it's ok to not document private methods if there's nothing worth to be mentioned

File: lib/pwnlib/tubes/serialtube.rb [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+      # than +@timeout+ seconds. It is ok for it to return some data in less
+      # time.
+      #
+      # @param [Integer] numbytes
+      #   An upper limit on the number of bytes to get.
+      #
+      # @return [String]
+      #   A string containing read bytes.
+      #
+      # @!macro raise_eof
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: Yard doesn't support cross-file macro definition naturally
https://stackoverflow.com/questions/10324926/how-to-make-yard-macros-apply-to-multiple-files

File: lib/pwnlib/tubes/tube.rb [hunk 13]
```diff
@@ -51,6 +51,9 @@ def initialize(timeout: nil)
       # @return [String]
       #   A string contains bytes received from the tube, or +''+ if a timeout occurred while
       #   waiting.
+      #
+      # @!macro raise_eof
+      # @!macro raise_timeout
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: Same as above

File: lib/pwnlib/tubes/tube.rb (hunk 2 above)
Comment (david942j) [reply lib/pwnlib/tubes/tube.rb]: So please rollback it

File: lib/pwnlib/tubes/tube.rb (hunk 3 above)
Comment (david942j) [reply lib/pwnlib/tubes/tube.rb]: But recv_raw is a block call 
The situation you mentioned should never happen

File: lib/pwnlib/tubes/tube.rb (hunk 13 above)
Comment (JonathanBeverley) [reply lib/pwnlib/tubes/tube.rb]: This one isn't cross-file.

Comment (david942j) [reply lib/pwnlib/tubes/tube.rb]: my fault

File: lib/pwnlib/tubes/tube.rb [hunk 14]
```diff
@@ -105,9 +108,13 @@ def recvpred(timeout: nil)
               data << c
             end
             data.slice!(0..-1)
-          ensure
+          # rubocop:disable Style/RescueStandardError
```
Comment (david942j) [inline lib/pwnlib/tubes/tube.rb]: And this must be changed back to ensure

File: lib/pwnlib/tubes/serialtube.rb [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    #
+    class SerialTube < Tube
+      # Instantiate a {Pwnlib::Tubes::SerialTube} object.
+      #
+      # @param [String] port
+      #   A device name for rubyserial to open, e.g. /dev/ttypUSB0
+      # @param [Integer] baudrate
+      #   Baud rate.
+      # @param [Boolean] convert_newlines
+      #   If +true+, will convert local +context.newline+ to +"\\r\\n"+ for remote
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: I guess no escape needed here?
`+"\r\n"+` should be enough

Comment (david942j) [reply lib/pwnlib/tubes/serialtube.rb]: And this sentence seems not complete

File: lib/pwnlib/tubes/serialtube.rb [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+        @serial_timer.countdown do
+          data = ''
+          begin
+            while @serial_timer.active?
+              data += @conn.read(numbytes - data.length)
+              break unless data.empty?
+              sleep 0.1
+            end
+            # TODO: should we reverse @convert_newlines here?
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: TODO(username)

File: lib/pwnlib/tubes/tube.rb (hunk 14 above)
Comment (JonathanBeverley) [reply lib/pwnlib/tubes/tube.rb]: I think I finally get what I misunderstood about this code, but I'll raise it as an issue so it can be discussed properly.

File: lib/pwnlib/tubes/serialtube.rb (hunk 15 above)
Comment (JonathanBeverley) [reply lib/pwnlib/tubes/serialtube.rb]: What I try it without the double escape, in the .html files generated by YARD, I see
```If true, will MARKER convert local context.newline to "rn" for remote```
I copied the syntax from tube.rb, which does the same thing. Note that in ``@example`` blocks the rules are different.

File: lib/pwnlib/tubes/serialtube.rb [hunk 17]
```diff
@@ -0,0 +1,111 @@
+# encoding: ASCII-8BIT
+
+require 'rubyserial'
+
+require 'pwnlib/tubes/tube'
+
+module Pwnlib
+  module Tubes
+    # Serial Connections
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: change here to
```ruby
# @!macro [new] raise_eof
#   @raise [Pwnlib::Errors::EndOfTubeError]
#     If the request is not satisfied when all data is received.

# Serial Connections
class SerialTube < Tube
```
otherwise YARD will have wrong result

File: lib/pwnlib/tubes/serialtube.rb [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+          begin
+            while @serial_timer.active?
+              data += @conn.read(numbytes - data.length)
+              break unless data.empty?
+              sleep 0.1
+            end
+            # XXX(JonathanBeverey): should we reverse @convert_newlines here?
+            return data
+          rescue RubySerial::Error
+            close
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: Current tests don't cover this rescue
add tests and ensure 100% coverage

File: lib/pwnlib/tubes/serialtube.rb [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+      #
+      # @!macro raise_eof
+      def send_raw(data)
+        raise ::Pwnlib::Errors::EndOfTubeError if @conn.nil?
+
+        data.gsub!(context.newline, "\r\n") if @convert_newlines
+        begin
+          return @conn.write(data)
+        rescue RubySerial::Error
+          close
```
Comment (david942j) [inline lib/pwnlib/tubes/serialtube.rb]: same here

Comment (JonathanBeverley) [reply lib/pwnlib/tubes/serialtube.rb]: done.

File: lib/pwnlib/tubes/serialtube.rb (hunk 18 above)
Comment (JonathanBeverley) [reply lib/pwnlib/tubes/serialtube.rb]: I can't see any way to do this. open(2) checks for all the same errors as read(2). Nothing I've done will cause an error in the right place, without writing bad code. 

If you know a way the test code can break Ruby's encapsulation model and directly invoke SerialTube.conn.close(), that could do it. But I really don't want to write a method to allow that, and honestly, if I did, I would then tighten up the error checking around @conn and close off that avenue as well.

Comment (JonathanBeverley) [reply lib/pwnlib/tubes/serialtube.rb]: I found a way.

**Comment by david942j** [comment]
Seems the tests sometimes hang forever
Any idea?

**Comment by JonathanBeverley** [comment]
I've tried chasing it down, but I can't reproduce it locally. It's probably happening because we're stuck in the Open3.popen3() block. Before I got the ensure working properly something similar would happen when an exception tried to break out of that block, but popen3 would wait for socat to complete before allowing it.

**Comment by david942j** [comment]
With `while true; do bundle exec rake; done` I can reproduce the hangs forever case.
It stucks on `Serial#read`, i.e. the `read` syscall

With quickly review I know the `rubyserial` gem do open files with NONBLOCK flag, so this should not happen.
Trying to figure out why.

**Comment by david942j** [comment]
Got it.
The main reason is a race condition between `rubyserial` and socat,
they both need to set (different) attributes on the same tty device.
Just add one line to make sure `socat` have finished its setup fix the issue: https://github.com/peter50216/pwntools-ruby/pull/121/commits/764ada530a3110f0044a04c26f03d18fccfc522d#diff-8dfbabae43dfe4e708d49fc1f86db83dR36

Now dealing with another fail case.. https://travis-ci.org/peter50216/pwntools-ruby/jobs/419754587

**Comment by david942j** [comment]
@JonathanBeverley I fixed the issues
While I don't have permission to push those fixes to the forked repo, could you push them to your branch?

The fixes contained in two commits: https://github.com/peter50216/pwntools-ruby/pull/121/commits/bda6f8bf57b15af7ac35ef66e8e861ebbaed2ebe and https://github.com/peter50216/pwntools-ruby/pull/121/commits/764ada530a3110f0044a04c26f03d18fccfc522d (plz ignore the debug messages)

**Comment by JonathanBeverley** [comment]
done.

**Comment by david942j** [comment]
Sorry those commits contains debug code..
please checked-out the change in `test/asm_test.rb` and `travis/install.sh`
